COSI-25, COSI-32: Implement Delete Bucket API, Add Tests, and Enhance CI

anurag4DSB opened the pull request:
Note for reviewers: Reviewing commit by commit is easier, as we have a commit story.

### Summary
This PR implements the DeleteBucket API and adds comprehensive tests and CI improvements:

Changes:
1. COSI-25: Implement Delete Bucket API:
- Deletes buckets only if they can be deleted with a simple operation (e.g., fails if objects exist).
- Ensures safe and fail-proof deletion handling.
2. COSI-25: Add Unit Tests:
- Added unit tests for successful deletion, error propagation, and edge cases.
3. COSI-32: Add e2e Tests:
- Integrated e2e tests for DeleteBucket into the CI pipeline to validate end-to-end functionality.
4. COSI-32: Ignore Cleanup on Failure:
- Skips cleanup steps after test failures for better error handling in CI.
5. COSI-25: Refactor Tests:
- Simplified test setup with helper functions and improved readability and maintainability.

Impact:
	•	Improves DeleteBucket reliability and test coverage.
	•	Enhances CI workflows and test maintainability.

Tickets: COSI-25, COSI-32

BourgoisMickael commented on .github/scripts/e2e_tests.sh:
```diff
@@ -239,4 +239,62 @@ if [[ "$USER_EXISTS" != *"NoSuchEntity"* ]]; then
+
+log_and_run echo "Verifying bucket deletion with name '$BUCKET_TO_BE_DELETED'..."
+
+for ((i=1; i<=$ATTEMPTS; i++)); do
+  BUCKET_HEAD_RESULT=$(aws --endpoint-url "$S3_ENDPOINT" s3api head-bucket --bucket "$BUCKET_TO_BE_DELETED" 2>&1 || true)
+
+  if [[ "$BUCKET_HEAD_RESULT" == *"Not Found"* ]]; then
+    log_and_run echo "Bucket with name '$BUCKET_TO_BE_DELETED' not found. Bucket deletion successful."
+    break
+  else
+    log_and_run echo "Attempt $i: Bucket with name '$BUCKET_TO_BE_DELETED' still exists. Retrying in $DELAY seconds..."
+    sleep $DELAY
+  fi
+done
+
+if [[ "$BUCKET_HEAD_RESULT" != *"Not Found"* ]]; then
+  log_and_run echo "Bucket with name '$BUCKET_TO_BE_DELETED' was not deleted after $ATTEMPTS attempts."
+  exit 1
+fi
```
The cli has a command to wait for a condition with retries: `aws s3api wait ...`

https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3api/wait/bucket-not-exists.html

And you can configure the retries with `max_attempts` and `retry_mode`: https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#retry-configuration

anurag4DSB replied:
hank you for the suggestion.

Unfortunately, the list-buckets API does not support this functionality. Since we only have the prefix Bucket_CLASS_NAME and not the exact bucket name, it cannot be used in this case.
- --retry and --delay: These options are not natively supported by the aws s3api list-buckets command. You would need to implement a custom retry logic, as demonstrated in the script above.
- aws s3api wait: This command is ideal for waiting on conditions such as bucket existence. However, it requires an exact bucket name and does not work with prefixes.

That being said, I have added support for retry logic in other cases, such as checking if a bucket is being deleted (when we know the exact bucket name) and verifying a user is being deleted (when we know the user’s name).

[Commit for bucket delete check](https://github.com/scality/cosi-driver/pull/61/commits/f6e5c90a25d5108136a22275dcb14756e23bc732)

[Commit for user delete check](https://github.com/scality/cosi-driver/pull/61/commits/13941bad39f39b785822394aea8b952d03ae2a12)

anurag4DSB replied:
Apologies for multiple comments, had some CI issues
FInally adding the passing commit link here for reference: https://github.com/scality/cosi-driver/pull/61/commits/7ff62b31272d4f8bc89da66c3fc1784179183771

BourgoisMickael commented on cosi-examples/bucketclaim-deletion-policy.yaml:
```diff
@@ -0,0 +1,8 @@
+spec:
+  bucketClassName: delete-bucket-class
+  protocols:
+  - s3
```
```suggestion
    - s3
```

indent

anurag4DSB replied:
its up for debate, I have usually seen (most common) without indent when using `-`.
I would prefer to follow the industry norm here, unless we have a specific standard or reason.

Technically both are correct though.

jonathan-gramain commented on .github/scripts/e2e_tests.sh:
```diff
@@ -239,4 +239,62 @@ if [[ "$USER_EXISTS" != *"NoSuchEntity"* ]]; then
+  if [[ "$BUCKET_HEAD_RESULT" == *"Not Found"* ]]; then
+    log_and_run echo "Bucket with name '$BUCKET_TO_BE_DELETED' not found. Bucket deletion successful."
+    break
+  else
```
You could get rid of the `else`

anurag4DSB replied:
changed, but probably will remove it based on @BourgoisMickael's comment above

anurag4DSB commented on pkg/driver/provisioner_server_impl.go:
```diff
@@ -256,7 +256,35 @@ func fetchS3Parameters(secretData map[string][]byte) (*util.StorageClientParamet
+	klog.V(3).InfoS("Received DriverDeleteBucket request", "bucketName", bucketName)
+	bucket, err := s.BucketClientset.ObjectstorageV1alpha1().Buckets().Get(ctx, bucketName, metav1.GetOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Failed to get bucket object from kubernetes", "bucketName", bucketName)
+		return nil, status.Error(codes.Internal, "failed to get bucket object from kubernetes")
```
Not focusing on logs too much right now, as we will change the logging mechanism next sprint. We will use OTel SDK for logs, which will probably remove all these logs

I added your comment in [this ticket](https://scality.atlassian.net/browse/COSI-35)

anurag4DSB commented:
rebased to main and pushed, didn't change anything yet
Now I am gonna address review comments

anurag4DSB commented on pkg/driver/provisioner_server_impl.go:
```diff
@@ -256,7 +256,35 @@ func fetchS3Parameters(secretData map[string][]byte) (*util.StorageClientParamet
+
+	s3Client, ok := client.(*s3client.S3Client)
+	if !ok {
+		klog.ErrorS(nil, "Unsupported client type for bucket creation", "bucketName", bucketName)
```
Thanks changed

anurag4DSB commented on pkg/driver/provisioner_server_impl.go:
```diff
@@ -256,7 +256,35 @@ func fetchS3Parameters(secretData map[string][]byte) (*util.StorageClientParamet
+		return nil, status.Error(codes.Internal, "failed to get bucket object from kubernetes")
+	}
+	parameters := bucket.Spec.Parameters
+	client, _, err := InitializeClient(ctx, s.Clientset, parameters, "S3")
```
Changed, thanks

anurag4DSB commented on pkg/clients/s3/s3_client.go:
```diff
@@ -85,3 +86,13 @@ func (client *S3Client) CreateBucket(ctx context.Context, bucketName string, par
+	if err != nil {
+		return err
+	}
+	return nil
```
Change, thanks
I indeed do not plan extra activities on this.

anurag4DSB commented:
Update: Rebased, and squashed commits
Also changed the refactor commit message to add the context of using AWS CLI retries and delays instead of for loop to check user and bucket deletion. We can use them in these cases as we know the respective bucket and user name, when we know the prefix we still use a for loop.

I will merge the PR.
